File: regression_tests/atry_struct.c
extern int __VERIFIER_nondet_int();

struct struct_of_struct {
    int x, y, z, t; 
    struct {
        int anon;
    } inner;
    struct {
        int anon;
    };
};

struct point {
    int x, y, z, t; 
    float weight;
} p78 = {1,2,3,4,4.004}, p79 = {1};    //p78 is fully initialised, p79 only has its first field set to 1, the rest is 0

struct point2 {
    int x, y, z, t;
    float weight;
} p1 = {1, 2, 3, 4, 67.0}, p2;           //all members of p1 and p2 to 0 by default because they are globals
struct point3;  //forward declaration
struct point3 {
    int x, y, z, t;
    float weight;
};
struct {        //anonymous declaration has no tag
    int x, y, z, t;
    float weight;
} anon1 = {1, 2, 3, 4, 67.0}, anon2;

int a = 1;

typedef struct plot plot;   //typedef of a forward declaration
struct plot {
    int x, y;
};
plot p = {10, 20};  // Works due to typedef


typedef struct {
    int n;
}S; //S is a typedef to an anonymous struct

S Sx;       //an object of type struct
/*
S Ax[5];    //an array of struct

int marks[10];

typedef struct Address {
    char city[50];
    int zip;
} Address;
typedef struct Person {
    char name[50];
    int age;
    Address address;  // No need for 'struct Address'
} Person;

//Person p = {"Alice", 25, {"Wonderland", 12345} };   //inner initialiser are string to be handled later

typedef struct {
    int x, y;
} Point;    //upper case P does not interfere with other anonymous struct point
typedef Point PointArray[10];  // PointArray is an array of 10 Points
PointArray points = {{1, 2}, {3, 4}};  // Array of points: only the first 2 points are initialised


typedef struct Node Node;   //Node is a typedef to struct Node which is a forward declaration
struct Node {               //the struct Node is declared
    int value;
    Node *next;             // Pointer to the same struct type using the earlier typedef
};
Node n = {10, 0};
*/

int main(void) {
    anon2 = anon1;
    struct point2 p3 = p1;
    int i = 0;
    struct point_1 { 
      int x, y, z, t; 
      float weight;
    } p78_1 = {1,2,3,4,4.004}, p79_1 = {1};    //p78_1 is fully initialised, p79_1 only has its first field set to 1, the rest is 'uninitiased'
    int gy;
    gy = p79_1.y;           //gy is uninitialised
    if (gy == __VERIFIER_nondet_int()) i--; 
    i++;
    if (p78_1.x == 1 && p79_1.x == 1 && p78.x == 1 && p79.x == 1) i++;  //always true
    else i--;                       //unreachable
    if (p79.y == 0) i++;            //always true
    else return 1;                  //unreachable
    float fl = p78.x;  //check that casting is called (int to float)
    int gh = p.x + p.y;
    if (gh == __VERIFIER_nondet_int()) i++;
    else i--;
    S x;
    x.n = __VERIFIER_nondet_int();
    if (x.n == 99) i++;
    else i--;
   struct point s1;
    struct point s2 = {1, 2, 3, 4, 42.42};  //list of initialisers can only be used at declaration time
    struct point s3 = {1, 2, 3};    //only partial initialisations is allowed: remaining fields are initialised to 0
    s1.x = __VERIFIER_nondet_int();
    if (s1.x == -5) i++;
    else i--;
    return 0;
}

File: atry_struct.pl
prolog_c([
declaration([extern, int], [function(UC___VERIFIER_nondet_int, [])]), 

declaration([struct(Struct_of_struct, [struct_decl([int], [x, y, z, t]), struct_decl([struct('anonymous', [struct_decl([int], [anon])])], [inner]), anonymous_member(struct('anonymous', [struct_decl([int], [anon])]))])]), 

declaration([struct(Point, [struct_decl([int], [x, y, z, t]), struct_decl([float], [weight])])], [initialised(P78, initializer([int(1), int(2), int(3), int(4), double(4.004)])), initialised(P79, initializer([int(1)]))]), 

declaration([struct(Point2, [struct_decl([int], [x, y, z, t]), struct_decl([float], [weight])])], [initialised(P1, initializer([int(1), int(2), int(3), int(4), double(67.0)])), P2]), 

declaration([struct(Point3)]), 

declaration([struct(Point3, [struct_decl([int], [x, y, z, t]), struct_decl([float], [weight])])]), 

declaration([struct('anonymous', [struct_decl([int], [x, y, z, t]), struct_decl([float], [weight])])], [initialised(Anon1, initializer([int(1), int(2), int(3), int(4), double(67.0)])), Anon2]), 

declaration([int], [initialised(A, int(1))]), 

declaration([typedef, struct(Plot)], [Plot]), 

declaration([struct(Plot, [struct_decl([int], [x, y])])]), 

declaration([Plot], [initialised(P, initializer([int(10), int(20)]))]), 

declaration([typedef, struct('anonymous', [struct_decl([int], [n])])], [UC_S]), 

declaration([UC_S], [UC_Sx]), 
function([int], function(Main, [param_no_decl([void], [])]), [], 
cmp_stmts([
stmt(assign(Anon2, Anon1)), 
declaration([int], [initialised(P3, P1)]), 
declaration([int], [initialised(I, int(0))]), 
declaration([struct(Point_1, [struct_decl([int], [x, y, z, t]), struct_decl([float], [weight])])], [initialised(P78_1, initializer([int(1), int(2), int(3), int(4), double(4.004)])), initialised(P79_1, initializer([int(1)]))]), 
declaration([int], [Gy]), 
stmt(assign(Gy, select(P79_1, y))), 
if_stmt(branch(1, equal_op(Gy, function_call(UC___VERIFIER_nondet_int, []))), 
stmt(postfix_dec_op(I)) ), 
stmt(postfix_inc_op(I)), 
if_stmt(branch(2, and_op(and_op(and_op(equal_op(select(P78_1, x), int(1)), equal_op(select(P79_1, x), int(1))), equal_op(select(P78, x), int(1))), equal_op(select(P79, x), int(1)))), 
stmt(postfix_inc_op(I)) , 
stmt(postfix_dec_op(I))), 
if_stmt(branch(3, equal_op(select(P79, y), int(0))), 
stmt(postfix_inc_op(I)) , 
return_stmt(int(1))
), 
declaration([float], [initialised(Fl, select(P78, x))]), 
declaration([int], [initialised(Gh, plus_op(select(P, x), select(P, y)))]), 
if_stmt(branch(4, equal_op(Gh, function_call(UC___VERIFIER_nondet_int, []))), 
stmt(postfix_inc_op(I)) , 
stmt(postfix_dec_op(I))), 
declaration([UC_S], [X]), 
stmt(assign(select(X, n), function_call(UC___VERIFIER_nondet_int, []))), 
if_stmt(branch(5, equal_op(select(X, n), int(99))), 
stmt(postfix_inc_op(I)) , 
stmt(postfix_dec_op(I))), 
declaration([struct(Point)], [S1]), 
declaration([struct(Point)], [initialised(S2, initializer([int(1), int(2), int(3), int(4), double(42.42)]))]), 
declaration([struct(Point)], [initialised(S3, initializer([int(1), int(2), int(3)]))]), 
stmt(assign(select(S1, x), function_call(UC___VERIFIER_nondet_int, []))), 
if_stmt(branch(6, equal_op(select(S1, x), minus_op(int(5)))), 
stmt(postfix_inc_op(I)) , 
stmt(postfix_dec_op(I))), 
return_stmt(int(0))

]))
]).